shell/server/pre-meteor.js

```diff
@@ -341,180 +341,208 @@ function serveStaticAsset(req, res) {
+          });
+        }
+      }).then((handler) => {
+        handler(req, res, (err) => {
+          if (err) {
+            next(err);
+          } else {
+            res.writeHead(404, { 'Content-Type': 'text/plain' });
+            res.end('404 not found: ' + req.url);
+          }
+        });
+      });
+    } else {
+      return Promise.resolve(undefined);
+    }
+  }).catch((err) => {
+    writeErrorResponse(res, err);
+  });
+};
 
-    // Retain the protocol & port from ROOT_URL but use the inbound
-    // hostname.
-    targetUrl.host = hostname + ':' + targetUrl.port;
+const redirectToMeteorOrServeStaticPublishing = (req, res, next) => {
+  if (!req.headers.host) {
+    res.writeHead(400, { 'Content-Type': 'text/plain' });
+    res.end('Missing Host header');
+    return;
+  }
 
-    // The following allows to avoid decoding + re-encoding query
-    // string parameters, if provided.
-    targetUrl = Url.resolve(Url.format(targetUrl), url);
-    return targetUrl;
-  };
+  return dispatchToMeteorOrStaticPublishing(req, res, next, true);
+};
 
-  const redirectToMeteorOrServeStaticPublishing = (req, res, next) => {
-    return dispatchToMeteorOrStaticPublishing(req, res, next, true);
-  };
+// This function serves responses on Sandstorm's main HTTP/HTTPS
+// port.
+const serveMeteorOrStaticPublishing = (req, res, next) => {
+  return dispatchToMeteorOrStaticPublishing(req, res, next, false);
+};
 
+// "Alternate ports" are ports other than the main HTTP or HTTPS
+// port. For requests to the shell & grains, we redirect to the main
+// port. For static publishing, we serve it.
+//
+// They are bound to FD #5 and higher.
+const getNumberOfAlternatePorts = function () {
+  const numPorts = process.env.PORT.split(',').length;
+  const numAlternatePorts = numPorts - 1;
+  return numAlternatePorts;
+};
+
+const listenOnAlternatePorts = function () {
   for (let i = 0; i < getNumberOfAlternatePorts(); i++) {
     // Call createServerForSandstorm() to skip our monkey patching.
     const alternatePortServer = Http.createServerForSandstorm(redirectToMeteorOrServeStaticPublishing);
     alternatePortServer.listen({ fd: i + 5 });
   }
+};
+
+Meteor.startup(() => {
+  const meteorRequestListeners = WebApp.httpServer.listeners("request");
 
-  const dispatchToMeteorOrStaticPublishing = (req, res, next, redirectRatherThanServeShell) => {
+  // Construct the middleware chain for requests to non-DDP, non-shell hosts.
+  const nonMeteorRequestHandler = Connect();
+  // BlackrockPayments is only defined in the Blackrock build of Sandstorm.
+  if (global.BlackrockPayments) { // Have to check with global, because it could be undefined.
+    nonMeteorRequestHandler.use(BlackrockPayments.makeConnectHandler(globalDb));
+  }
+
+  nonMeteorRequestHandler.use(serveMeteorOrStaticPublishing);
+
+  WebApp.httpServer.removeAllListeners("request");
+  WebApp.httpServer.on("request", (req, res) => {
```

Comment:
I'm still a bit nervous about the fact that this stretch of code does not have an exception handler. Sure, maybe we can verify by inspection now that nothing bad is likely to happen, but who knows how the code will evolve? Especially because this code gets hit on *every* request, it seems like going out of our way for robustness would be worthwhile.